Sources/Foundation/JSONEncoder.swift

```diff
@@ -417,570 +646,338 @@ fileprivate struct _JSONKeyedEncodingContainer<K : CodingKey> : KeyedEncodingCon
-        self.codingPath.append(_JSONKey(index: self.count))
-        defer { self.codingPath.removeLast() }
+    mutating func nestedContainer<NestedKey>(keyedBy _: NestedKey.Type, forKey key: Self.Key) ->
+        KeyedEncodingContainer<NestedKey> where NestedKey: CodingKey
+    {
+        let convertedKey = self._converted(key)
+        let newPath = self.codingPath + [convertedKey]
+        let object = self.object.setObject(for: convertedKey.stringValue)
+        let nestedContainer = JSONKeyedEncodingContainer<NestedKey>(impl: impl, object: object, codingPath: newPath)
+        return KeyedEncodingContainer(nestedContainer)
+    }
 
-        let dictionary = NSMutableDictionary()
-        self.container.add(dictionary)
+    mutating func nestedUnkeyedContainer(forKey key: Self.Key) -> UnkeyedEncodingContainer {
+        let convertedKey = self._converted(key)
+        let newPath = self.codingPath + [convertedKey]
+        let array = self.object.setArray(for: convertedKey.stringValue)
+        let nestedContainer = JSONUnkeyedEncodingContainer(impl: impl, array: array, codingPath: newPath)
+        return nestedContainer
+    }
 
-        let container = _JSONKeyedEncodingContainer<NestedKey>(referencing: self.encoder, codingPath: self.codingPath, wrapping: dictionary)
-        return KeyedEncodingContainer(container)
+    mutating func superEncoder() -> Encoder {
+        let newEncoder = self.getEncoder(for: _JSONKey.super)
+        self.object.set(newEncoder, for: _JSONKey.super.stringValue)
+        return newEncoder
     }
 
-    public mutating func nestedUnkeyedContainer() -> UnkeyedEncodingContainer {
-        self.codingPath.append(_JSONKey(index: self.count))
-        defer { self.codingPath.removeLast() }
-        
-        let array = NSMutableArray()
-        self.container.add(array)
-        return _JSONUnkeyedEncodingContainer(referencing: self.encoder, codingPath: self.codingPath, wrapping: array)
+    mutating func superEncoder(forKey key: Self.Key) -> Encoder {
+        let convertedKey = self._converted(key)
+        let newEncoder = self.getEncoder(for: convertedKey)
+        self.object.set(newEncoder, for: convertedKey.stringValue)
+        return newEncoder
     }
+}
 
-    public mutating func superEncoder() -> Encoder {
-        return _JSONReferencingEncoder(referencing: self.encoder, at: self.container.count, wrapping: self.container)
+extension JSONKeyedEncodingContainer {
+    @inline(__always) private mutating func encodeFloatingPoint<F: FloatingPoint & CustomStringConvertible>(_ float: F, key: CodingKey) throws {
+        let value = try self.wrapFloat(float, for: key)
+        self.object.set(value, for: key.stringValue)
+    }
+    
+    @inline(__always) private mutating func encodeFixedWidthInteger<N: FixedWidthInteger>(_ value: N, key: CodingKey) throws {
+        self.object.set(.number(value.description), for: key.stringValue)
```

Comment:
All of the callers of this function are already expanded out per type, could this line not just be used at each call size instead?

Reply:
If it is okay for you, I'd prefer to keep it this way. Since Swift does not support Macros this is the closest to Macros we can get (`@inline(__always)`). In case we ever need to change something in the encoding method for `Int`s we only need to change it at one point.